pytorch_lightning/strategies/collaborative.py

```diff
@@ -0,0 +1,504 @@
+                "Enabling delay_optimizer_step, delay_state_averaging or offload_optimizer "
+                "requires a scheduler_fn to be passed to the strategy if a scheduler is being used "
+                "(this is because the optimizer is re-created within Hivemind)."
+            )
+
+        scheduler = self._scheduler_fn if self._require_scheduler_fn else None
+        params = optimizer.param_groups if self._require_scheduler_fn else None
+        optimizer = type(optimizer) if self._require_scheduler_fn else optimizer
+        opt = hivemind.Optimizer(
+            dht=self.dht,
+            run_id=self._run_id,
+            params=params,
+            optimizer=optimizer,
+            scheduler=scheduler,
+            target_batch_size=self._target_batch_size,
+            batch_size_per_step=self._batch_size,
+            **self._optimizer_kwargs,
+        )
+
+        if not self._scheduler_fn:
+            self._wrap_schedulers(opt)
+        opt.load_state_from_peers()
+        self.optimizers = [opt]
+        self._opt = opt
+
+        if self._reuse_grad_buffers:
+            assert self.lightning_module is not None
+            # turn off zero_grad by Lightning
+            if is_overridden("optimizer_zero_grad", self.lightning_module):
+                rank_zero_warn(
+                    "You have overridden `optimizer_zero_grad` which will be disabled. "
+                    "When CollaborativeStrategy(reuse_grad_buffers=True), "
+                    "the optimizer cannot call zero grad, as this would "
+                    "delete the gradients before they are averaged."
+                )
+
+            def override_fn(*args: Any, **kwargs: Any) -> None:
+                pass
+
+            self.lightning_module.optimizer_zero_grad = override_fn  # type: ignore[assignment]
+
+    def _wrap_schedulers(self, opt: "hivemind.Optimizer") -> None:
+        # wrap schedulers so that they only update when the hivemind optimizer updates
+        for scheduler_config in self.lr_scheduler_configs:
+            scheduler_config.scheduler = HiveMindScheduler(
+                scheduler=scheduler_config.scheduler,
+                optimizer=opt,
+            )
+
+    def on_train_batch_start(self, batch: Any, batch_idx: int, dataloader_idx: int = 0) -> None:
+        if not self._hivemind_initialized:
+            self._hivemind_initialized = True
+            if self._batch_size is None:
+                try:
+                    self._batch_size = extract_batch_size(batch)
+                    log.info(f"Found per machine batch size automatically from the batch: {self._batch_size}")
+                except Exception as e:
+                    raise MisconfigurationException(
+                        "We tried to infer the batch size from the first batch of data are were unable to. "
+                        "Please provide the batch size to the Strategy (CollaborativeStrategy"
+                        f"exception raised: {e}"
```

Comment:
This is the error message inside `extract_batch_size`

```
    error_msg = (
        "We could not infer the batch_size from the batch. Either simplify its structure"
        " or provide the batch_size as `self.log(..., batch_size=batch_size)`."
    )
```

So, there would be a duplication and confusion with both the errors being mentioned.

Reply:
Since we catch the underlying exception, this will never be printed to the stdout

I've modified this exception to be more specific (I was just catching Exception beforehand)